### PR TITLE
pyperclip upgrade

### PR DESCRIPTION
**Pull request by krey**
- [x] closes https://github.com/pandas-dev/pandas/issues/22707
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/clipboard/__init__.py
Comment (jbrockmendel) [inline pandas/io/clipboard/__init__.py]: is this needed?

**Comment by jbrockmendel** [comment]
are there any tests to go with this?

File: pandas/io/clipboard/__init__.py
Comment (krey) [reply pandas/io/clipboard/__init__.py]: What do you mean by *this*?

I only altered the original pyperclip 1.7.0 code as much as it was necessary

**Comment by krey** [comment]
> are there any tests to go with this?

We could add pyperclip's tests to pandas'. 

I'm not sure of specific was to test https://github.com/pandas-dev/pandas/issues/22707

File: pandas/io/clipboard/__init__.py
Comment (jbrockmendel) [reply pandas/io/clipboard/__init__.py]: i mistook this comment for commented-out code, my mistake.

**Comment by jbrockmendel** [comment]
Looks like CI is failing because an Exception subclass was removed `from pandas.io.clipboard.exceptions import PyperclipException`

**Comment by WillAyd** [comment]
What's the reasoning behind moving everything out of submodules into one `__init__`? Also can you explain the PR you are undoing from the OP? Is that a revived package now?

**Comment by krey** [comment]
> What's the reasoning behind moving everything out of submodules into one `__init__`?

I don't know why they changed `pyperclip` that way.

> Also can you explain the PR you are undoing from the OP?

I don't mind doing a manual merge if needed, but I wasn't sure if it mattered.

> Is that a revived package now?

Not to my knowledge.

**Comment by krey** [comment]
Honestly I'm not even sure why `__init__.py` isn't just

```
from pyperclip import copy, paste
clipboard_get = paste
clipboard_set = copy
``` 
or something to that effect.

Is it to reduce the number of dependencies?

Or purely historical reasons?

**Comment by WillAyd** [comment]
Have to excuse my ignorance on this part of the codebase - I wasn't even aware we vendored this :-) OK I think I'm on board with this, since it's just an update to 1.7.0.

Can you check if this also solves #11720 ? cc @datapythonista as well for #26887

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you also add a whatsnew for v1.0.0?

**Comment by datapythonista** [comment]
I'm not sure if the clipboard tests are running in the CI, so we need to be very careful here.  I fixed it in #26949, but it made the tests be much slower, and the PR wasn't finally merged. Not sure if anyone else worked on it, but it may be that tests for the clipboard are not running at all, so we can't merge this without fixing that or proper "manual" testing.

Also, I guess there was a reason to copy the dependency to pandas, but may be things changed and now we can simply have it as a dependency?

**Comment by krey** [comment]
If there's no opposition to changing `pyperclip` to a dependency, I will close this PR and open a new one.

@WillAyd I don't have access to a Windows machine (re https://github.com/pandas-dev/pandas/issues/11720)

**Comment by WillAyd** [comment]
> If there's no opposition to changing `pyperclip` to a dependency

I don't think we should do that. If starting from scratch might make sense but I don't see the churn being worth it. These aren't a huge burden to keep included

**Comment by datapythonista** [comment]
I see the advantage of getting updates automatically, not a risky PR like this for every new version of pyperclip. And I don't see a disadvantage, I don't think a PR to make it a dependency is a big change, is it?

**Comment by krey** [comment]
@datapythonista Agreed.

The change is literally https://github.com/pandas-dev/pandas/pull/28471#issuecomment-532017034 (and changing some imports in the tests)

**Comment by TomAugspurger** [comment]
Like Will, I'd prefer to keep vendoring it. I don't think it's worth adding
as a dependency.

**Review by jreback — CHANGES_REQUESTED** [review]
this is fine to vendor and update to current; we don't want to add actual required dependencies (and making this optional is actually somewhat non-trivial).

pls add a whatsnew note on the issue resolution. do we need a test for this issue?

**Review by datapythonista — CHANGES_REQUESTED** [review]
I think I found out the problem that made tests very slow when trying to restore the clipboard tests (afaik they are not yet working).

Once #28482 is merged, I'll give it a try. Probably worth waiting here until we have the clipboard tests back, to make sure this upgrade is not breaking anything.

**Comment by WillAyd** [comment]
@krey looks like the above fix is in. Can you merge master and address comments around the whatsnew / test?

**Comment by krey** [comment]
Shall we add pyperclip's tests?https://github.com/asweigart/pyperclip/blob/master/tests/test_pyperclip.py

**Comment by WillAyd** [comment]
I think that makes sense to include as part of this

**Comment by WillAyd** [comment]
@krey can you fix merge conflicts and pull in the tests?

File: pandas/tests/io/test_pyperclip.py
Comment (WillAyd) [inline pandas/tests/io/test_pyperclip.py]: Hmm we don't use unittest so would need to change these - can you see how much effort it is to convert to pytest idiom?

Trying to balance asking a lot here versus getting this to work; unfortunately our current tests for this are suspect

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by datapythonista** [comment]
Do we want to run pyperclip tests? I understand that we don't want to add pyperclip as a hard dependency, and it's too complex to have it as an optional dependency for now. But I think we just want to have it as a vendored dependency, trusting that they run their own tests, and I don't see which value adds to pandas to run their tests (or include them in the project).

What are we going to do if tests fail in our CI? I don't think we want to touch our copy of their source, do we?

Am I missing something?

**Comment by WillAyd** [comment]
I'm not tied to it. I guess I'm unclear as to how we are currently validating / testing these though. Is this back up and running in CI? Or something we need to do as a pre-cursor?

Just hesitant to blanket vendor these changes without some kind of CI validation

**Comment by jreback** [comment]
we don’t need the tests; straight vendor is enough

**Comment by datapythonista** [comment]
pyperclip itself should be tested. And we should have tests for our functions that use it, so if the new version breaks something in our code, the CI should be failing. So I'm with Jeff, we don't need the tests.

**Comment by WillAyd** [comment]
OK reverted just reverted that for @krey let's see

**Comment by jreback** [comment]
@krey this looks good. can you merge master and I think have a linting issue in the whatsnew. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by krey** [comment]
@jreback It's mostly green...

**Comment by WillAyd** [comment]
codecov is sometimes wonky so no worries there; rest lgtm as well

**Review by WillAyd — APPROVED** [review]

File: pandas/io/clipboard/__init__.py
Comment (simonjayhawkins) [inline pandas/io/clipboard/__init__.py]: mypy error: Not all arguments converted during string formatting  [str-format]

**Pull request merged**
